engine/src/main/scala/cromwell/engine/workflow/lifecycle/finalization/CopyWorkflowMetadataActor.scala

```diff
@@ -0,0 +1,120 @@
+import cromwell.services.metadata.impl.builder.MetadataBuilderActor.{BuiltMetadataResponse, ReadyToBuildResponse}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.{Duration, FiniteDuration}
+import scala.util.{Failure, Success}
+
+object CopyWorkflowMetadataActor {
+  def props(workflowId: WorkflowId, ioActor: ActorRef, serviceRegistryActor: ActorRef, workflowDescriptor: EngineWorkflowDescriptor, workflowOutputs: CallOutputs,
+            initializationData: AllBackendInitializationData) = Props(
+    new CopyWorkflowMetadataActor(workflowId, ioActor, serviceRegistryActor, workflowDescriptor, workflowOutputs, initializationData)
+  ).withDispatcher(IoDispatcher)
+}
+
+sealed trait CopyWorkflowMetadataActorState
+
+case object Initial extends CopyWorkflowMetadataActorState
+case object WaitingState extends CopyWorkflowMetadataActorState
+case object GetMetadata extends CopyWorkflowMetadataActorState
+
+case class CopyWorkflowMetadataActorData(
+                                          metaDataPath: String,
+                                          builderActor: Option[ActorRef],
+                                          respActor: Option[ActorRef]
+                                        )
+
+class CopyWorkflowMetadataActor(workflowId: WorkflowId, override val ioActor: ActorRef, serviceRegistryActor: ActorRef, val workflowDescriptor: EngineWorkflowDescriptor, workflowOutputs: CallOutputs,
+                                initializationData: AllBackendInitializationData)
+  extends LoggingFSM[CopyWorkflowMetadataActorState, Option[CopyWorkflowMetadataActorData]] with MetadataServicesStore with MetadataDatabaseAccess
+    with PathFactory with AsyncIoActorClient {
+  implicit val ec = context.dispatcher
+  override lazy val ioCommandBuilder = GcsBatchCommandBuilder
+  override val pathBuilders = workflowDescriptor.pathBuilders
+
+  private val metadataReadTimeout: Duration = Duration(30, "seconds")
+
+  def readMetadataWorkerActorProps(): Props = ReadDatabaseMetadataWorkerActor.props(metadataReadTimeout).withDispatcher(ServiceDispatcher)
+
+  startWith(Initial, None)
+
+  when(Initial) {
+    case Event(Finalize, _) =>
+      val respondTo: ActorRef = sender
+      val selv = context.self
+      workflowDescriptor.getWorkflowOption(FinalWorkflowMetadataDir) match {
+        case Some(metadataPath) =>
+          val mba = context.actorOf(MetadataBuilderActor.props(readMetadataWorkerActorProps))
+          mba ! SwitchToWaitMetadata(selv)
+          goto(WaitingState) using Option(CopyWorkflowMetadataActorData(metadataPath, Option(mba), Option(respondTo)))
+        case None =>
+          respondTo ! FinalizationSuccess
+          stay()
+      }
+  }
+
+  when(WaitingState) {
+    case Event(ReadyToBuildResponse, data) => {
+      val query = MetadataQuery(workflowId, None, None, None, None, false)
+      val timeout = FiniteDuration(30, "seconds")
+      val mba = data.get.builderActor.get
```

Comment:
Also, it looks like you are doing `.get` from an `Option` two times in a row. Someone can put `None` instead of `Some(...)` and your code will throw an exception.
You should refactor it somehow to avoid such problems.